libraries/shared/logs/delegator.go

```diff
@@ -0,0 +1,85 @@
+// VulcanizeDB
+// Copyright © 2019 Vulcanize
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logs
+
+import (
+	"errors"
+	"github.com/sirupsen/logrus"
+	"github.com/vulcanize/vulcanizedb/libraries/shared/chunker"
+	"github.com/vulcanize/vulcanizedb/libraries/shared/transformer"
+	"github.com/vulcanize/vulcanizedb/pkg/core"
+	"github.com/vulcanize/vulcanizedb/pkg/datastore"
+)
+
+var ErrNoTransformers = errors.New("no event transformers configured in the log delegator")
+
+type ILogDelegator interface {
+	AddTransformer(t transformer.EventTransformer)
+	DelegateLogs(errs chan error, logsFound chan bool)
+}
+
+type LogDelegator struct {
+	Chunker       chunker.Chunker
+	LogRepository datastore.HeaderSyncLogRepository
+	Transformers  []transformer.EventTransformer
+}
+
+func (delegator *LogDelegator) AddTransformer(t transformer.EventTransformer) {
+	delegator.Transformers = append(delegator.Transformers, t)
+	delegator.Chunker.AddConfig(t.GetConfig())
+}
+
+func (delegator *LogDelegator) DelegateLogs(errs chan error, logsFound chan bool) {
+	if len(delegator.Transformers) < 1 {
+		errs <- ErrNoTransformers
+		return
+	}
+
+	persistedLogs, fetchErr := delegator.LogRepository.GetUntransformedHeaderSyncLogs()
+	if fetchErr != nil {
+		logrus.Errorf("error loading logs from db: %s", fetchErr.Error())
+		errs <- fetchErr
```

Comment:
So I may be missing something here but I think we need to pass errors through a channel if we want to execute the delegator as a goroutine. I know we definitely want to execute the extractor as a goroutine so it doesn't block the delegator while it's syncing logs for perhaps 1000s of headers, but maybe we could get away with executing the delegator synchronously?

Reply:
Yeah, that makes sense. How we handle these errors up the call stack: do we fatally error on the first occurrence, or could we retry a few times? Since the processes are now disconnected, one could assume the transformations have some food to chew while we retry a bit to get some more resilience to DB/network load :>

Reply:
I think we currently log fatal, but I'm on board with just logging the error and retrying 👍